docs/web-vitals.md

```diff
@@ -0,0 +1,362 @@
+ * browser receives the first byte of the response (a.k.a. TTFB).
+ */
+"fcp.time_to_first_byte": number;
+
+/**
+ * The delta between TTFB and the first contentful paint (FCP).
+ */
+"fcp.time_since_first_byte": number;
+
+/**
+ * The loading state of the document at the time when FCP `occurred (see
+ * `LoadState` for details). Ideally, documents can paint before they finish
+ * loading (e.g. the `loading` or `dom-interactive` phases).
+ */
+"fcp.load_state": LoadState;
+
+```
+
+The event time is equal to the start of the page load, the duration is equal to `fcp.value`
+
+### [Time to First Byte](https://web.dev/articles/ttfb) attributes
+
+```js
+/** Shared fields */
+"web_vital.name": "TTFB",
+"web_vital.rating": "good" | "needs-improvement" | "poor",
+"web_vital.id": string;
+
+/** TTFB specific fields */
+
+"ttfb.value": number;
+"ttfb.delta": number;
+
+/**
+ * The total time from when the user initiates loading the page to when the
+ * DNS lookup begins. This includes redirects, service worker startup, and
+ * HTTP cache lookup times.
+ */
+"ttfb.waiting_time": number;
+/**
+ * The total time to resolve the DNS for the current request.
+ */
+"ttfb.dns_time": number;
+/**
+ * The total time to create the connection to the requested domain.
+ */
+"ttfb.connection_time": number;
+/**
+ * The time time from when the request was sent until the first byte of the
+ * response was received. This includes network time as well as server
+ * processing time.
+ */
+"ttfb.request_time": number;
+
+```
+
+The event time is equal to the timeOrigin & the duration is equal to ttfb.value
+
+## Customization of event attributes
+
+For more fine-tuned event processing, pass in a custom callback functions for each of the web-vitals:
+
+```js
+
+const sdk = new HoneycombWebSDK({
+  webVitals: {
+    onCLS: (clsWithAttribution) => {//custom event processor },
+  }
+});
+
+```
+
+Supported fields:
+
+```ts
+onCLS?: CLSReportCallbackWithAttribution;
+onFCP?: FCPReportCallbackWithAttribution;
+onFID?: FIDReportCallbackWithAttribution;
+onINP?: INPReportCallbackWithAttribution;
+onLCP?: LCPReportCallbackWithAttribution;
+onTTFB?: TTFBReportCallbackWithAttribution;
+
+```
+
+The callbacks are passed their respective metric with attribution. [web-vitals docs](https://github.com/GoogleChrome/web-vitals?tab=readme-ov-file#metricwithattribution)
+
+**Note**: passing in a custom callback will _replace_ the default attributes provided by this package, not augment.
+
+## Additional Configuration
+
+### Using custom data attribute to identify elements
+
+Sometimes the element selector (used by LCP, INP & CLS events) is not very human-readable (or changes frequently) & makes it difficult to figure out the actual culprit. Pass in a custom data selector & hny will use that identifier instead for the `lcp.element`, `cls.element` & `inp.element` attributes.
+
+```js
+const sdk = new HoneycombWebSDK({
+  serviceName: 'my-app',
+  webVitals: {
```

Comment:
Maybe this API should be `webVitalsInstrumentation`? And this is where we can also enable/disable it? The other thought I had was keeping it consistent with how other auto-instrumentation is configured which would be:

```js
  instrumentations: [new WebVitalsInstrumentation({
    enabled: true,
    elementDataAttribute: 'human-id'
  })]
```

The upside of this is keeping it consistent with other auto-instrumentation and not having to maintain another layer between our API and the instrumentation. The downside is having to import and inititalize `new WebVitalsInstrumentation({})` if the user wants any customization or to disable the instrumentation.

Reply:
👍  I like the adjustment of `webVitalsInstrumentation`.

🤔  I also like the second suggestion that makes it closer to upstream auto-instrumentation setup... but acknowledge the potential yuck factor with requiring someone to do that extra work to customize. If this instrumentation does get added upstream though, it's likely to have this type of interface. So is it better to go with that now? 

Also I guess I'm curious about how much data this generates and maybe we need a PM input on how common this will need to be adjusted, and if you're at the point where you want to make those customizations, then step-by-step instructions regarding adding another package maybe isn't terrible?

Reply:
i think for me it hinges on whether or not we want to ship web-vital support out-of-the-box, by default. 

if yes, i'm inclined towards 

```js
const sdk = new HoneycombWebSDK({
  serviceName: 'my-app',
  webVitalsInstrumentation: {//...custom config},
})

```

When the config is absent, we rig our default set-up. if false, opt out. 

but if we don't want a "web-vitals, batteries-included" version of the wrapper, 
```
instrumentations: [new WebVitalsInstrumentation({
    enabled: true,
    elementDataAttribute: 'human-id'
  })]
```

definitely makes more sense (and we could probably scrape the `enabled` option altogether)